halo2-ecc/src/fields/native_fp.rs

```diff
@@ -0,0 +1,213 @@
+        self.range
+    }
+    fn limb_bits(&self) -> usize {
+        F::CAPACITY as usize
+    }
+
+    fn get_assigned_value(&self, x: &AssignedValue<F>) -> F {
+        *x.value()
+    }
+
+    fn load_private(&self, ctx: &mut Context<F>, a: F) -> AssignedValue<F> {
+        ctx.load_witness(a)
+    }
+
+    fn load_constant(&self, ctx: &mut Context<F>, a: F) -> AssignedValue<F> {
+        ctx.load_constant(a)
+    }
+
+    // signed overflow BigInt functions
+    fn add_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        b: impl Into<AssignedValue<F>>,
+    ) -> AssignedValue<F> {
+        self.gate().add(ctx, a.into(), b.into())
+    }
+
+    fn add_constant_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        c: F,
+    ) -> AssignedValue<F> {
+        let c = self.load_constant(ctx, c);
+        self.add_no_carry(ctx, a, c)
+    }
+
+    fn sub_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        b: impl Into<AssignedValue<F>>,
+    ) -> AssignedValue<F> {
+        self.gate().sub(ctx, a.into(), b.into())
+    }
+
+    // Input: a
+    // Output: p - a if a != 0, else a
+    fn negate(&self, ctx: &mut Context<F>, a: AssignedValue<F>) -> AssignedValue<F> {
+        self.gate().neg(ctx, a)
+    }
+
+    fn scalar_mul_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        c: i64,
+    ) -> AssignedValue<F> {
+        let c_f = if c >= 0 {
+            let c_abs = u64::try_from(c).unwrap();
+            F::from(c_abs)
+        } else {
+            let c_abs = u64::try_from(-c).unwrap();
+            -F::from(c_abs)
+        };
+
+        let c_loaded = ctx.load_constant(c_f);
+        self.gate().mul(ctx, a.into(), c_loaded)
+    }
+
+    fn scalar_mul_and_add_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        b: impl Into<AssignedValue<F>>,
+        c: i64,
+    ) -> AssignedValue<F> {
+        let a_mul_c = self.scalar_mul_no_carry(ctx, a, c);
+        self.add_no_carry(ctx, a_mul_c, b)
+    }
+
+    fn mul_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        b: impl Into<AssignedValue<F>>,
+    ) -> AssignedValue<F> {
+        self.gate().mul(ctx, a.into(), b.into())
+    }
+
+    // noop
+    fn check_carry_mod_to_zero(&self, _ctx: &mut Context<F>, _a: AssignedValue<F>) {}
+
+    // noop
+    fn carry_mod(&self, _ctx: &mut Context<F>, a: AssignedValue<F>) -> AssignedValue<F> {
+        a
+    }
+
+    fn range_check(
+        &self,
+        _ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        max_bits: usize, // the maximum bits that a.value could take
+    ) {
+        let a: AssignedValue<F> = a.into();
+        debug_assert!(a.value().to_bytes_le().len() <= max_bits);
```

Comment:
you need to do a real `range_check` here!

Reply:
i guess an optimization is if `max_bits >= F::NUM_BITS`, then you can skip the range check